yjit/src/codegen.rs

```diff
@@ -1292,39 +1292,63 @@ fn gen_newrange(
 }
 
 fn guard_object_is_heap(
+    ctx: &mut Context,
     asm: &mut Assembler,
-    object_opnd: Opnd,
+    object: Opnd,
+    object_opnd: YARVOpnd,
     side_exit: Target,
 ) {
+    let object_type = ctx.get_opnd_type(object_opnd);
+    if object_type.is_heap() {
+        return;
+    }
+
     asm.comment("guard object is heap");
 
     // Test that the object is not an immediate
-    asm.test(object_opnd, (RUBY_IMMEDIATE_MASK as u64).into());
+    asm.test(object, (RUBY_IMMEDIATE_MASK as u64).into());
     asm.jnz(side_exit);
 
     // Test that the object is not false
-    asm.cmp(object_opnd, Qfalse.into());
+    asm.cmp(object, Qfalse.into());
     asm.je(side_exit);
+
+    if object_type.diff(Type::UnknownHeap) != usize::MAX {
+        ctx.upgrade_opnd_type(object_opnd, Type::UnknownHeap);
+    }
```

Comment:
This is saying "upgrade to UnknownHeap unless object_type is incompatible with UnknownHeap" right? And it doesn't try to upgrade when we already know that `object_type` is an immediate, for example.

Off topic for this change, I think the return value of Context::diff should be changed to an enum so callers don't need to go-to-definition to find out that usize::MAX means incompatible.

Reply:
> This is saying "upgrade to UnknownHeap unless object_type is incompatible with UnknownHeap" right? And it doesn't try to upgrade when we already know that object_type is an immediate, for example. Callers probably should make sure to CantCompile if they already know that the guard is going to fail

I agree that we might be doing something wrong somewhere. The previous version of this PR was using `asm.jmp(side_exit)` when `diff == usize::MAX`, which I believe is not that different from `CantCompile`, but somehow it significantly reduced `ratio_in_yjit` on railsbench. It wasn't immediately obvious why it happened, so I kept the generated code almost as is, leaving the investigation for a future PR.

> but this check gives caller the flexibility to not worry about failing an assert when they don't check at compile time.

I've wanted to make `CantCompile` a Rust `Err`. It would make a caller as simple as `guard_object_is_heap(...)?;`. Clearly it's not for this PR, but I can work on it if you agree with the idea.

> Off topic for this change, I think the return value of Context::diff should be changed to an enum so callers don't need to go-to-definition to find out that usize::MAX means incompatible.

:+1: